From ae867867ae6c8268efd4c1c80144eabed0342337 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 15 Feb 2014 10:06:21 +0100
Subject: [PATCH] Corrected bug in gnutls_pcert_list_import_x509_raw().

The bug caused gnutls_pcert_list_import_x509_raw() to crash if
gnutls_x509_crt_list_import() would fail with the provided data.
Reported by Dmitriy Anisimkov.
---
 lib/gnutls_pcert.c |    9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/lib/gnutls_pcert.c b/lib/gnutls_pcert.c
index 56c2d6f..344b3e9 100644
--- a/lib/gnutls_pcert.c
+++ b/lib/gnutls_pcert.c
@@ -124,7 +124,7 @@ gnutls_pcert_list_import_x509_raw(gnutls_pcert_st * pcerts,
 					flags);
 	if (ret < 0) {
 		ret = gnutls_assert_val(ret);
-		goto cleanup;
+		goto cleanup_crt;
 	}
 
 	for (i = 0; i < *pcert_max; i++) {
@@ -138,14 +138,15 @@ gnutls_pcert_list_import_x509_raw(gnutls_pcert_st * pcerts,
 	ret = 0;
 	goto cleanup;
 
-      cleanup_pcert:
+ cleanup_pcert:
 	for (j = 0; j < i; j++)
 		gnutls_pcert_deinit(&pcerts[j]);
 
-      cleanup:
+ cleanup:
 	for (i = 0; i < *pcert_max; i++)
 		gnutls_x509_crt_deinit(crt[i]);
-
+ 
+ cleanup_crt:
 	gnutls_free(crt);
 	return ret;
 
-- 
1.7.10.4

